# Profitable App Profiles for the App Store and Google Play Markets

#### My goal in this project is to analyze apps that are free but also profitable from in-app advertisements. This analyzation consists of app data from both the Google Play and Apple App Store markets.

Free app revenue is determined by the number of active users of that app. I want to analyze the data and determine what type of apps are profitable to help developers understand the apps categories that are likely to attract the most users.

##### Full data can be found:

As of 2018 there were 2 million iOS apps on the App Store and 2.1 million on the Google Play Store. To avoid the time and cost of collecting data on 4 million+ apps, we will use sample data from 2 separate data sets:

- A [dataset](https://www.kaggle.com/datasets/lava18/google-play-store-apps) containing data about approximately 10,000 Android apps from Google Play;the data was collected in August 2018

- A [dataset](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps) containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017



#  Exploring our dataset

I will start by opening the 2 data sets to prepare for exploration:

In [2]:
from csv import reader

# Google Play
opened_file = open('googleplaystore.csv', encoding="utf")
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[:]

# App Store
opened_file = open('AppleStore.csv',  encoding="utf")
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[:]

- I'll write a function to explore the 2 data sets to be used further on in our analysis.


In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line between rows
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [4]:
print(android_header)
print('\n')
explore_data(android, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10842
Number of columns: 13


- Print the column names, and try to identify the columns that could help us with our analysis.
Let's examine the App Store data set:

In [5]:
del(ios_header[0])
print(ios_header)
print('\n')
explore_data(ios, 0, 3, True)

['track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 7198
Number of columns: 15


##### Nga leximi i te dhenave rezulton se ne 'Apple Store' ka gjithsej 7197 rreshta dhe 16 kolona, nga te cilat me te rendesishme per qellimin tone jane id, track_name, currency, price, user_rating etj.
##### Ndersa nga te dhenat e 'Google Play Store' rezulton se ka gjithesej 10842 rreshta dhe 13 kolona, nga te cilat ato me te rendesishme per punen tone jane app, rating, reviews, size, price etj.
The Google Play data set has 10,841 rows and 13 columns. Browsing the column names, the ones that appear to be most useful for our analysis are App, Category, Reviews, Installs, Type, Price, and Genres

##   Preprocessing the Data

#### Data cleaning

- We do data cleaning before the analysis; it includes removing or correcting wrong data, removing duplicate data, and modifying the data to fit the purpose of our analysis.

- Detect inaccurate data, and correct or remove it.

The data set for the Google Play apps has a discussion session. A user posted that there is a missing entry for 'Category' on row 10,472. We will print that row below to see if the user is correct.

In [6]:
print(android[10472])
print(android_header)

['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


- There is no category for row 10,472. We will delete this row and then print the length of the data set to cofirm that the row is gone. This command should only be run once

In [7]:
del android[10472]
print(len(android))


10841


- In the iOS data set, the first column is the index for the row. This is not useful to our analysis. I will remove the index column from the data set:

In [8]:
for app in ios:
    del(app[0])
explore_data(ios, 0, 3, True)

['size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 7198
Number of columns: 14


## Duplicate Entries
#### Some apps in the Google Play data set have more than one entry. For example, if we search for Instagram there are 4 rows for a single app:
- Detect duplicate data, and remove the duplicates.

In [9]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


- There are actually 1,181 duplicate apps in the data set:

In [10]:
duplicates = []
unique = []

for app in android:
    name = app[0]
    if name in unique:
        duplicates.append(name)
    else:
        unique.append(name)
print('Number of duplicates: ', len(duplicates))
print('\n')
print('Examples of duplicates: ', duplicates[:10])

Number of duplicates:  1181


Examples of duplicates:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


### Here we have write the code to remove the duplicate entries.

- Examing the duplicate entries for Instagram above shows us the main difference between the entries is in the 4th column which is the number of reviews. Rather than remove entries at random, we'll keep the row with the highest number of reviews as that row will give us the more reliable data.

- To achieve this we will create a dictionary consisting of the app name and the max reviews of that app.


In [11]:
max_reviews = {}

for app in android:
    name = app[0]
    n_reviews = app[3]
    
    if name in max_reviews and max_reviews[name] < n_reviews:
        max_reviews[name] = n_reviews
    elif name not in max_reviews:
        max_reviews[name] = n_reviews

In [12]:
print('Expected length: ', len(android) - 1181)
print('Actual length: ', len(max_reviews))

Expected length:  9660
Actual length:  9660


We can now use the max_reviews dictionary to remove the duplicate rows throughout the data set. We'll only keep the entries that contain the highest number of reviews.

- Create 2 lists, android_clean and already_added
- Loop through the app rows
- For each row, assign the variable 'name' to the name of the app
- Convert the number of reviews to a float and assign to n_reviews
- If n_reviews equals the same amount as the max_reviews and name is not alreay in the already_added list, append the row to  android_clean
- Append the app name to already_added to keep track of the already added apps

In [14]:
# we have created new list for the non-duplicated data
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = app[3]
    if (n_reviews == max_reviews[name]) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)
# We will quickly explore our android_clean list to confirm the number of rows is 9,569
explore_data(android_clean, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 9660
Number of columns: 13


# Removing Non-English Apps
- We'd like to only analyze apps developed for English speaking audiences. Both the data sets contain apps developed for non-English speaking audiences. We should remove the apps that contain non-English titles. The best way to do this is to look for any non-English characters in the app name and compare that to the ASCII system. Any character greater than an ASCII value of 127 should reveal that the app name has a non-English title.

- We can use the built in Python function ord() to get the corresponding encoded value for a character in a string. We will loop through each title and check the ASCII value of each character.

In [15]:
def check_english(str):
    for char in str:
        if ord(char) > 127:
            return False
    return True

print(check_english('Instagram'))
print(check_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
False


- Some apps in the data set are English but have non-English characters in them such as emoji's and specialized characters. To minimize data loss, we will only remove the app from the data set if it contains 3 or more non-English ASCII characters

In [16]:
def check_english(str):
    special_chars = 0
    for char in str:
        if ord(char) > 127:
            special_chars += 1
    if special_chars > 3:
        return False
    else:
        return True

print(check_english('Docs To Go™ Free Office Suite'))
print(check_english('Instachat 😜'))

True
True


- Below we will use the check_english() function to filter out non-English apps for both Google and iOS data sets:

In [17]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if check_english(name):
        android_english.append(app)
for app in ios:
    name = app[1]
    if check_english(name):
        ios_english.append(app)

explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 9615
Number of columns: 13


['size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '

### Isolating the free apps
- Our goal with this analysis is to identify the trends related to free apps that have in-app adds. We need to remove any apps in the data sets that have a cost to download:

In [18]:
android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
for app in ios_english:
    price = app[4]
    if price == '0':
        ios_final.append(app)
      
explore_data(android_final, 0, 3, True)
print('\n')
explore_data(ios_final, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 8861
Number of columns: 13


['NAVIRO(ナビロー) - カーナビ/バイクナビ/徒歩ナビが使える高性能ナビアプリ', '55756800', 'USD', '0.0', '0', '0', '0.0', '0.0', '2.2.9', '4+', 'Navigation', '37', '5', '1', '1']


['Confusions de lettres', '5000192', 'USD', '2.99', '0', '0', '0.0', '0.0', '1.0.4', '4+', 'Education', '38', '5', '1', '1']


['最長１週間の献立が簡単に作れるme:new（ミーニュー）', '39937024', 'USD', '0.0', '0', '0', '0.0', '0.0', '1,2.4', '4+', 'Food & Drink', '38', 

- There are 8,864 Android apps and 3,222 iOS apps left to analyze. These are the cleaned data sets consisting of non-duplicate, English developed, and free apps

# Common Apps by Genre
- My aim is to find the apps that attract the most amount of users. Free apps that have in-app advertisements are dependent upon amount of users to be profitable. My validation strategy for an app idea will have 3 steps:

- Build a minimal Android version of the app, and add it to Google Play.
If the app has a good response from users, we then develop it further.
If the app is profitable after six months, we also build an iOS version of the app and add it to the App Store.
We need to find apps that are successful in both Android and Apple markets. I'll begin the analysis by generating a frequency table for the prime_genre column in the iOS data and Genres and Category columns of the Google Play data.

I'll build 2 functions:

- One to generate frequency tables that show percentages
- One to display the tables in desc order

In [19]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    table_percentage = {}
    for key in table:
        percent = (table[key] / total) * 100
        table_percentage[key] = percent
    return table_percentage

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

print(display_table(ios_final, 11))

37 : 39.07427341227126
38 : 31.86221743810549
40 : 23.896663078579117
43 : 3.229278794402583
24 : 0.6458557588805167
25 : 0.4305705059203444
35 : 0.32292787944025836
26 : 0.32292787944025836
39 : 0.1076426264800861
36 : 0.1076426264800861
None


Of the free English apps in the iOS data, more than 58% are Games. The next closest genre is Entertainment with close to 8%, followed by Photo & Video with almost 5%.

The App Store is dominated by apps developed for fun, not as many for practical use. But just because these apps have the highest frequencies does not imply that they have as many users.

We'll continue our analysis with the Google Play data set:

In [20]:
print(display_table(android_final, 1)) #Category

FAMILY : 18.936914569461685
GAME : 9.694165444080802
TOOLS : 8.452770567656021
BUSINESS : 4.593161042771697
LIFESTYLE : 3.9047511567543167
PRODUCTIVITY : 3.8934657487868187
FINANCE : 3.7016138133393524
MEDICAL : 3.521047285859384
SPORTS : 3.3969077982169056
PERSONALIZATION : 3.3066245344769216
COMMUNICATION : 3.2389120866719328
HEALTH_AND_FITNESS : 3.080916375126961
PHOTOGRAPHY : 2.9454914795169844
NEWS_AND_MAGAZINES : 2.7987811759395105
SOCIAL : 2.663356280329534
TRAVEL_AND_LOCAL : 2.3360794492720913
SHOPPING : 2.245796185532107
BOOKS_AND_REFERENCE : 2.144227513824625
DATING : 1.8620923146371742
VIDEO_PLAYERS : 1.794379866832186
MAPS_AND_NAVIGATION : 1.3993905879697552
FOOD_AND_DRINK : 1.2413948764247829
EDUCATION : 1.1736824286197944
ENTERTAINMENT : 0.9592596772373322
LIBRARIES_AND_DEMO : 0.9366888613023362
AUTO_AND_VEHICLES : 0.9254034533348381
HOUSE_AND_HOME : 0.8238347816273558
WEATHER : 0.8012639656923597
EVENTS : 0.7109807019523756
PARENTING : 0.6545536621148855
ART_AND_DESIGN :

- Practical applications appear to be better represented in the Google Play store. We can confirm this further by examing the frequency table for the Genres column


In [21]:
print(display_table(android_final, 9)) #Category


Tools : 8.441485159688522
Entertainment : 6.071549486513937
Education : 5.349283376594064
Business : 4.593161042771697
Productivity : 3.8934657487868187
Lifestyle : 3.8934657487868187
Finance : 3.7016138133393524
Medical : 3.521047285859384
Sports : 3.464620246021894
Personalization : 3.3066245344769216
Communication : 3.2389120866719328
Action : 3.103487191061957
Health & Fitness : 3.080916375126961
Photography : 2.9454914795169844
News & Magazines : 2.7987811759395105
Social : 2.663356280329534
Travel & Local : 2.3247940413045933
Shopping : 2.245796185532107
Books & Reference : 2.144227513824625
Simulation : 2.0426588421171425
Dating : 1.8620923146371742
Arcade : 1.8508069066696762
Video Players & Editors : 1.7718090508971898
Casual : 1.749238234962194
Maps & Navigation : 1.3993905879697552
Food & Drink : 1.2413948764247829
Puzzle : 1.1285407967498025
Racing : 0.9931159011398263
Role Playing : 0.9366888613023362
Libraries & Demo : 0.9366888613023362
Auto & Vehicles : 0.92540345333483

The Genres column is much more granular than the Category column. We don't need the granular genres for our analysis right now so we will stick with using Category going forward. The Apple Store is dominated by apps designed for fun and games, while the Google Play store has a blend of for-fun and practical apps. Next in our analysis will be to determine which tpyes of apps have the most users.

 # Most Popular Apps by Genre
The easiest way to determine popularity of an app is to calculate the average number of installs per category. We can use the Installs column for the Google Play data but unfortunately iOS does not have a similar column. We can use the number of user ratings as a proxy. This number can be found in the rating_count_tot column.

Below I'll calculate the average number of user ratings / installs per app on the App Store:

In [22]:
ios_genres = freq_table(ios_final, 11)

for genre in ios_genres:
    total = 0
    len_genre = 0
    
    for app in ios_final:
        genre_app = app[11]
        if genre_app == genre:
            total += float(app[5])
            len_genre += 1
    avg_rating = total / len_genre
    print(genre, ':', avg_rating)

37 : 0.0
38 : 0.0
40 : 0.0
43 : 0.0
25 : 0.0
24 : 0.0
35 : 0.0
26 : 0.0
36 : 0.0
39 : 0.0


- Navigation has the highest number of reviews, but that figure is being skewed by a few large apps like Google Maps and Waze:

In [23]:
for app in ios_final:
    if app[11] == 'Navigation':
        print(app[1], ':', app[5])

- Our goal is to find popular app genres but some of these apps may seem more poular than they are. Many categories such as Navigation, Social Networking, and Music are dominated by a few highly influential apps (Facebook, Twitter, Spotify etc.).

Reference apps have 74,942 ratings. A majority of these are coming from 2 large apps, the Bible and Dictionary.com

In [24]:
for app in ios_final:
    if app[11] == 'Reference':
        print(app[1], ':', app[5])

In [25]:
for app in ios_final:
    if app[11] == 'Productivity':
        print(app[1], ':', app[5])

- Let's review the number of installs by category on the Google Play Store. One thing to note about how this data set calculates install numbers is they are not exact, rather it is a tiered structure. See below:

In [26]:
display_table(android_final, 5) #installs column
android_categories = freq_table(android_final, 1)

1,000,000+ : 15.743144114659746
100,000+ : 11.544972350750479
10,000,000+ : 10.518000225708159
10,000+ : 10.202008802618215
1,000+ : 8.396343527818532
100+ : 6.91795508407629
5,000,000+ : 6.838957228303803
500,000+ : 5.574991535944025
50,000+ : 4.773727570251665
5,000+ : 4.51416318699921
10+ : 3.5436181017943795
500+ : 3.250197494639431
50,000,000+ : 2.290937817402099
100,000,000+ : 2.1216566978896285
50+ : 1.9185193544746644
5+ : 0.7899785577248618
1+ : 0.5078433585374111
500,000,000+ : 0.27084979121995256
1,000,000,000+ : 0.22570815934996052
0+ : 0.0451416318699921
0 : 0.011285407967498025
